Add source control instructions (#1187)
vaadin/docs: articles/flow/configuration/source-control.adoc

---
title: Source Control
description: Track and manage changes to the source code in a Vaadin project.
order: 130
---

= Source Control

In addition to the standard directory layout of typical Java applications, Vaadin projects also include a number of files and folders related to the frontend tooling.
A typical Vaadin application has a directory layout with something like the following content:

----
frontend              <1>
└── generated/

> └── themes/           <2>

Reviewer: I wonder if it would be clearer to show the actual theme directory inside `themes`?

Author: Yeah, I thought about that. It would be clearer but also make it more crowded.

Author: Added the themes substructure

> └── index.html

Reviewer 2: Could also say that `index.html` is auto-generated but can be customized (and only in the latter case would we recommend adding it to source control.)

Author: Done. I added a description of `index.html`, and added it to gitignore with a note.

node_modules/         <3>              
src/
└── main              <4>
    └── java 
    └── resources
└── test/             <5>
target/
package.json          <6>
package-lock.json     <7>
pom.xml               <8>
webpack.config.js     <9>
webpack.generated.js  <10>
tsconfig.json         <11>

> types.d.ts

Author: I don't know what is the description of `types.d.ts`.

Reviewer: It can contain additional type definitions used in the project.

Right now it defines that css files are imported as `CSSResultGroup` but that is not important to the user and it will also probably change at some point. It should in almost all Flow cases be considered auto generated on demand and gitignored, just like `tsconfig.json`

Author: Description added

----
<1> Frontend resources, including CSS, TypeScript, and JavaScript files are placed under this folder.
<2> This folder includes custom themes.
<3> A folder that caches the frontend modules that the project depends upon.
This folder is auto-generated based on the contents of [filename]#package.json# and [filename]#package-lock.json#.
<4> Application sources.
<5> Test sources.

> <6> Holds the metadata related to the project's frontend configuration and dependencies.

Reviewer: Maybe this could say "defines version ranges for frontend dependencies" and <7> would say "Defines the exact versions for frontend dependencies in this project" as "locks the frontend dependencies" is not super clear

Author: Done

<7> Locks the frontend dependencies and their sub-dependencies.

> <8> Project and configuration details used by Maven to build the project.

Reviewer: + Java dependencies?

Author: I'll leave the description on this one as it is. People should know what is a pom.xml. And if they don't, this is not going to be the page to teach them on the topic.

<9> Can optionally be used to customize webpack configuration.

> <10> An auto-generated file holding Flow's default webpack configuration.

Reviewer: ```suggestion
<10> An auto-generated file containing the webpack configuration needed for all applications.
```

> <11> An auto-generated file used to compile the TypeScript code in the project if Lit templates are used.

Reviewer: ```suggestion
<11> An auto-generated file that defines the configuration for  compiling TypeScript code in the project, when needed.
```

.Project content may vary
[NOTE]

> ====

Reviewer 3: Is an admonition really needed? I think the content would work fine just as a plain paragraph.

Author: I don't have a strong opinion on this, but I feel that an admonition is more easily noticeable as it highlights the content compared to the above description of the directory structure.

![image](https://user-images.githubusercontent.com/42799254/158992987-b07d45f0-81ea-4064-a09e-24472f0aca33.png)

Author: Do you have a strong opinion @jouni?

Reviewer 3: No, not a strong opinion. Only that in general, try to keep the number of admonitions low per article.

The aforementioned directory layout may vary depending on the project's configuration.
For example, a project using `pnpm` instead of `npm` will have a [filename]#pnpm-lock.yaml# file instead of [filename]#package-lock.json#.
====

The following [filename]#.gitignore# file lists the files and folders that should be excluded from a typical Vaadin project. 

.`.gitignore`
[source]
----
/target/
.gradle
build/
!**/src/main/**/target/
!**/src/test/**/target/
!**/src/main/**/build/
!**/src/test/**/build/

.DS_Store

# The following files are generated/updated by vaadin-maven-plugin or vaadin-gradle-plugin
node_modules/
frontend/generated/
pnpmfile.js
.npmrc
webpack.generated.js
vite.generated.ts

# Browser drivers for local integration tests
drivers/

# Error screenshots generated by TestBench for failed integration tests
error-screenshots/

# Eclipse and STS
.apt_generated
.classpath
.factorypath
.project
.settings
.springBeans
.sts4-cache

# IntelliJ IDEA
.idea
*.iws
*.iml
*.ipr
out/
!**/src/main/**/out/
!**/src/test/**/out/

# NetBeans
/nbproject/private/
/nbbuild/
/dist/
/nbdist/
/.nb-gradle/

# VS Code
.vscode/

# Maven wrapper
.mvn/
mvnw

> ----

Reviewer: I think `tsconfig.json` and `types.d.ts` could be in this list as they are hardly ever customized and newer Vaadin versions might bring changes. Maybe with a comment to remove from gitignore if you want to customize them

Author: done
